src/debug/debug_stream/debug_stream_thread_info.c

```diff
@@ -0,0 +1,371 @@
+
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <rtos/alloc.h>
+#include <ipc/topology.h>
+#include <stdio.h>
+#include <soc.h>
+#include <adsp_debug_window.h>
+
+#include <user/debug_stream_thread_info.h>
+
+LOG_MODULE_REGISTER(thread_info);
+
+#define THREAD_INFO_MAX_THREADS 64
+/* Must be bigger than sizeof(struct thread_info_record_hdr) */
+#define THREAD_INFO_INITIAL_RECORD_BUFFER_SIZE 256
+
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+/* Data structure to store the cycle counter values from the previous
+ * round. The numbers are used to calculate what the load was on this
+ * round.
+ */
+static struct previous_counters { /* Cached data from previous round */
+	uint64_t active;	  /* All execution cycles */
+	uint64_t all;		  /* All cycles including idle */
+	struct thread_counters {
+		void *tid;	 /* thread ID (the thread struct ptr) */
+		uint64_t cycles; /* cycle counter value */
+	} threads[THREAD_INFO_MAX_THREADS]; /* The max amount of threads we follow */
+} __aligned(CONFIG_DCACHE_LINE_SIZE) previous[CONFIG_MP_MAX_NUM_CPUS];
+#endif
+
+/*
+ * Buffer structure for building the Debug Stream record before
+ * sending it forward. The buffer is persistent and each record is
+ * build in the same buffer. Its reallocated as double size if it
+ * becomes too small.
+ */
+struct record_buf {
+	size_t size;
+	size_t w_ptr;
+	uint8_t *buf;
+};
+
+/*
+ * Data structure to be passed down to thread_info_cb() by
+ * k_thread_foreach_current_cpu().
+ */
+struct user_data {
+	int core;
+	struct record_buf *rec;
+	int thread_count;
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+	bool stats_valid;
+	uint32_t all_cycles;
+	void *active_threads[THREAD_INFO_MAX_THREADS];
+	struct previous_counters *previous;
+#endif
+};
+
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+
+/* Finds and/or updates cached cycle counter value for 'tid'-thread
+ * and calculates used execution cycles since previous round and
+ * returns it.
+ */
+
+static uint32_t thread_info_get_cycles(void *tid, k_thread_runtime_stats_t *thread_stats,
+				       struct user_data *ud, const char *name)
+
+{
+	int i;
+
+	/* Mark the thread as active */
+	ud->active_threads[ud->thread_count] = tid;
+	/* look for cached value from previous round for 'tid'-thread */
+	for (i = 0; i < ARRAY_SIZE(ud->previous->threads); i++) {
+		if (ud->previous->threads[i].tid == tid) {
+			/* Calculate number cycles since previous round */
+			uint32_t cycles = (uint32_t) (thread_stats->execution_cycles -
+						      ud->previous->threads[i].cycles);
+
+			LOG_DBG("%p found at %d (%s %llu)", tid, i,
+				name, thread_stats->execution_cycles);
+			/* update cached value */
+			ud->previous->threads[i].cycles = thread_stats->execution_cycles;
+			return cycles;
+		}
+	}
+
+	/* If no cached value was found, look for an empty slot to store the recent value */
+	for (i = 0; i < ARRAY_SIZE(ud->previous->threads); i++) {
+		if (ud->previous->threads[i].tid == NULL) {
+			ud->previous->threads[i].tid = tid;
```

Comment:
can be a follow-up, but maybe we can have a better name than `previous` here. You don't store more than one snapshot, so isn't it just "current"?

Reply:
It the cycles number from the previous round. After the calculations it will be "current", but when its used next time its again from "previous" round.